src/Microsoft.ManagedZlib/src/Deflater.cs

```diff
@@ -0,0 +1,273 @@
+        //Vivi's note> MemoryHandle was for managing the pointers. Gone now
+        private bool _isDisposed;
+        private const int minWindowBits = -15;  // WindowBits must be between -8..-15 to write no header, 8..15 for a
+        private const int maxWindowBits = 31;   // zlib header, or 24..31 for a GZip header
+
+        // Note, DeflateStream or the deflater do not try to be thread safe.
+        // The lock is just used to make writing to unmanaged structures atomic to make sure
+        // that they do not get inconsistent fields that may lead to an unmanaged memory violation.
+        // To prevent *managed* buffer corruption or other weird behaviour users need to synchronise
+        // on the stream explicitly.
+        private object SyncLock => this;
+
+        internal Deflater(CompressionLevel compressionLevel, int windowBits)
+        {
+            Debug.Assert(windowBits >= minWindowBits && windowBits <= maxWindowBits);
+            ManagedZLib.CompressionLevel zlibCompressionLevel;
+            int memLevel;
+
+            switch (compressionLevel)
+            {
+                // See the note in ManagedZLib.CompressionLevel for the recommended combinations.
+
+                case CompressionLevel.Optimal:
+                    zlibCompressionLevel = ManagedZLib.CompressionLevel.DefaultCompression;
+                    memLevel = ManagedZLib.Deflate_DefaultMemLevel;
+                    break;
+
+                case CompressionLevel.Fastest:
+                    zlibCompressionLevel = ManagedZLib.CompressionLevel.BestSpeed;
+                    memLevel = ManagedZLib.Deflate_DefaultMemLevel;
+                    break;
+
+                case CompressionLevel.NoCompression:
+                    zlibCompressionLevel = ManagedZLib.CompressionLevel.NoCompression;
+                    memLevel = ManagedZLib.Deflate_NoCompressionMemLevel;
+                    break;
+
+                case CompressionLevel.SmallestSize:
+                    zlibCompressionLevel = ManagedZLib.CompressionLevel.BestCompression;
+                    memLevel = ManagedZLib.Deflate_DefaultMemLevel;
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(compressionLevel));
+            }
+
+            ManagedZLib.CompressionStrategy strategy = ManagedZLib.CompressionStrategy.DefaultStrategy;
+
+            ZErrorCode errC;
+            try
+            {
+                errC = ManagedZLib.CreateZLibStreamForDeflate(out _zlibStream, zlibCompressionLevel,
+                                                             windowBits, memLevel, strategy);
+            }
+            catch (Exception cause)
+            {
+                throw new ZLibException("ZLibErrorDLLLoadError - The underlying compression routine could not be loaded correctly.", cause);
+            }
+
+            switch (errC)
+            {
+                case ZErrorCode.Ok:
```

Comment:
As @jkoritzinsky suggested already, you could remove all the ZErrorCode usages and simplify the code by just adding asserts. The error codes were used because they came from native code and we needed to capture them and throw something different depending on the value. Not anymore.